post: using Partytown 🎉 in Next.js (#54)
cloudx-labs/posts: posts/navarroaxel/10.improve-your-website-performance-using-patytown.md

---
title: Improve your website performance using Partytown type in script tags
published: true
description: Move third-party scripts without any alterations to web workers using Partytown in React with Next.js.
tags: 'javascript, webdev, performance, react'
cover_image: ./assets/using-partytown-in-next-cover.jpg
id: 1189226
---

> Can we use Google Tag Manager from a [web worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API)? Is `<script type="text/partytown">` a new standard? Let's check what’s this new `script` type and how easy can we move work load from the [main thread](https://developer.mozilla.org/en-US/docs/Glossary/Main_thread) to a worker with an example using GTM and Next.js. After that you can implement this with vanilla JS and other third party scripts like Facebook Pixel, Mixpanel, Amplitude, etc.

Author: ```suggestion
Can we use Google Tag Manager from a [web worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API)? Is `<script type="text/partytown">` a new standard? Let's check what’s this new `script` type and how easily can we move work load from the [main thread](https://developer.mozilla.org/en-US/docs/Glossary/Main_thread) to a worker with an example using GTM and Next.js. After that you can implement this with vanilla JS and other third party scripts like Facebook Pixel, Mixpanel, Amplitude, etc.
```

```html
<script type="text/partytown">
  /* GTM Script Here */
</script>
<script type="text/partytown" src="https://example.com/analytics.js"></script>
```

### Where is the `forward`?

To `forward` a function call to the worker we just need to add the configuration to the `window`  object before the inlined Partytown `script`.

```html
<script>
  window.partytown = { forward: ['dataLayer.push'] };
</script>
<script type="text/partytown">
  /* GTM script here */
</script>
```

Now, you can use `window.dataLayer.push()` from the main thread. 🙌

## Using workers in Next.js

Setting up Partytown in a Next.js website is veryeasy! We just need Next.js [v12.1.1](https://github.com/vercel/next.js/releases/tag/v12.1.1) or later.

🧠 Remember that this is experimental feature in Next.js and it’s not covered by semver. 🙀

First, we should enable the experimental flag in the `next.config.js` file.

```javascript
module.exports = {
  experimental: { nextScriptWorkers: true }
};
```

And install the Partytown package.

```bash
npm install --save-dev @builder.io/partytown
```

> The Nest.js workers don't require extra configurations, but we should `forward` the `dataLayer.push` for this example. We need to include this configuration in a `<Head>` component inside a custom `_document.js`.

Author: ```suggestion
The Next.js workers don't require extra configurations, but we should `forward` the `dataLayer.push` for this example. We need to include this configuration in a `<Head>` component inside a custom `_document.js`.
```

              lib: '/_next/static/~partytown/',
              debug: true,
              forward: ['dataLayer.push']
            };
          `
        }}
      />
    </Head>
    <body>
      <Main />
      <NextScript />
    </body>
  </Html>
);

export default Document;
```

💡 The `debug: true` setting gives us nice info in the DevTools' console.

Finally, we can add the GTM script in our `_app.js`.

```javascript
import Script from 'next/script'; 

const MyApp = ({ Component, pageProps }) => (
  <>
    <Script
      id="gtm-script"
      strategy="worker"
      dangerouslySetInnerHTML={{
        __html: `/* GTM script here */`
      }}
    />
    <Component {...pageProps} />
  </>
);

export default MyApp;
```

The `id` property is required for _inline scripts_ in order to let Next.js track and optimize the script.

And it's done! Your GTM runs in a worker that you can communicate with using the `dataLayer.push` - as always.

You can find installation instructions for [React](https://partytown.builder.io/react), [Angular](https://partytown.builder.io/angular), [Nuxt](https://partytown.builder.io/nuxt) and [more](https://partytown.builder.io/integrations).

## Conclusion

> Too much JavaScript can block DOM construction, delaying how quickly pages can render. Reducing JavaScript in our main thread is a good strategy to reduce the time until interaction. This will free up main thread resources to only be used for the primary web app execution, e.g. network requests.

Author: ```suggestion
Too much JavaScript can block DOM construction, delaying how quickly pages can render. Reducing JavaScript in our main thread is a good strategy to reduce the time until interaction. This will free up main thread resources so they can only be used for the primary web app execution, e.g. network requests.
```

---
title: Improve your website performance using Partytown type in script tags
published: true
description: Move third-party scripts without any alterations to web workers using Partytown in React with Next.js.
tags: 'javascript, webdev, performance, react'
cover_image: ./assets/using-partytown-in-next-cover.jpg
id: 1189226
---

> Can we use Google Tag Manager from a [web worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API)? Is `<script type="text/partytown">` a new standard? Let's check what’s this new `script` type and how easily can we move work load from the [main thread](https://developer.mozilla.org/en-US/docs/Glossary/Main_thread) to a worker with an example using GTM and Next.js. After that you can implement this with vanilla JS and other third party scripts like Facebook Pixel, Mixpanel, Amplitude, etc.

Reviewer: ```suggestion
Can we use Google Tag Manager from a [web worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API)? Is `<script type="text/partytown">` a new standard? Let's see what this new `script` type is and how we can easily move workloads from the [main thread](https://developer.mozilla.org/en-US/docs/Glossary/Main_thread) to a worker with an example using GTM and Next.js. After that you can implement this with vanilla JS and other third party scripts like Facebook Pixel, Mixpanel, Amplitude, etc.
```

![Without Partytown and With Partytown: Your code and third-party code compete for main thread resources](https://user-images.githubusercontent.com/452425/152393346-6f721a4f-3f66-410a-8878-a2b49e24307f.png)

### Creating script tags for workers

To run code in the worker just use `type="text/partytown"`. The browser won’t recognize that type ignoring the `script` tag and then Partytown will load that `script` inside a web worker. This works for inline code or files.

```html
<script type="text/partytown">
  /* GTM Script Here */
</script>
<script type="text/partytown" src="https://example.com/analytics.js"></script>
```

### Where is the `forward`?

To `forward` a function call to the worker we just need to add the configuration to the `window`  object before the inlined Partytown `script`.

```html
<script>
  window.partytown = { forward: ['dataLayer.push'] };
</script>
<script type="text/partytown">
  /* GTM script here */
</script>
```

Now, you can use `window.dataLayer.push()` from the main thread. 🙌

## Using workers in Next.js

> Setting up Partytown in a Next.js website is veryeasy! We just need Next.js [v12.1.1](https://github.com/vercel/next.js/releases/tag/v12.1.1) or later.

Reviewer: ```suggestion
Setting up Partytown in a Next.js website is very easy! We just need Next.js [v12.1.1](https://github.com/vercel/next.js/releases/tag/v12.1.1) or later.
```

> 🧠 Remember that this is experimental feature in Next.js and it’s not covered by semver. 🙀

Reviewer: ```suggestion
🧠 Remember that this is an experimental feature in Next.js and it’s not covered by semver. 🙀
```

              lib: '/_next/static/~partytown/',
              debug: true,
              forward: ['dataLayer.push']
            };
          `
        }}
      />
    </Head>
    <body>
      <Main />
      <NextScript />
    </body>
  </Html>
);

export default Document;
```

💡 The `debug: true` setting gives us nice info in the DevTools' console.

Finally, we can add the GTM script in our `_app.js`.

```javascript
import Script from 'next/script'; 

const MyApp = ({ Component, pageProps }) => (
  <>
    <Script
      id="gtm-script"
      strategy="worker"
      dangerouslySetInnerHTML={{
        __html: `/* GTM script here */`
      }}
    />
    <Component {...pageProps} />
  </>
);

export default MyApp;
```

The `id` property is required for _inline scripts_ in order to let Next.js track and optimize the script.

And it's done! Your GTM runs in a worker that you can communicate with using the `dataLayer.push` - as always.

You can find installation instructions for [React](https://partytown.builder.io/react), [Angular](https://partytown.builder.io/angular), [Nuxt](https://partytown.builder.io/nuxt) and [more](https://partytown.builder.io/integrations).

## Conclusion

> Too much JavaScript can block DOM construction, delaying how quickly pages can render. Reducing JavaScript in our main thread is a good strategy to reduce the time until interaction. This will free up main thread resources so they can only be used for the primary web app execution, e.g. network requests.

Reviewer: ```suggestion
Too much JavaScript can block DOM construction, delaying how quickly pages can render. Reducing JavaScript in our main thread is a good strategy to reduce the time until interaction. This will free up main thread resources so they can be used for the primary web app execution, e.g. network requests.
```
